### PR TITLE
[Runtime] Embed Maven 3.9.1

### PR DESCRIPTION
**Pull request by Eskibear**
See https://github.com/eclipse-m2e/m2e-core/issues/1169#issuecomment-1422023478

### CODE REVIEW

**Review by laeubi — CHANGES_REQUESTED** [review]
@Eskibear Currently Maven 3.9.x has some breaking changes, so before we use it we need to wait until most plugins have adopted to it or find a way to let the user choose the embedded maven version.

File: pom.xml [hunk 1]
```diff
@@ -102,7 +102,7 @@
 		<dependency>
 			<groupId>org.eclipse.m2e</groupId>
 			<artifactId>org.eclipse.m2e.maven.runtime</artifactId>
-			<version>3.8.700-SNAPSHOT</version>
+			<version>3.9.000-SNAPSHOT</version>
```
Comment (laeubi) [inline pom.xml:105]:
```suggestion
			<version>3.9.0-SNAPSHOT</version>
```

**Comment by Eskibear** [comment]
Thank for the info. I somehow managed to make it compile by https://github.com/eclipse-m2e/m2e-core/pull/1238/commits/b582ab6644778045d15ae1a77b8bea8c9822ea03 

> we need to wait until most plugins have adopted to it

Do we have a list of these plugin and status of their compatibility with 3.9.0? Let me know if there's anything I can help.

File: org.eclipse.m2e.mavenarchiver/src/org/eclipse/m2e/mavenarchiver/internal/AbstractMavenArchiverConfigurator.java [hunk 2]
```diff
@@ -466,7 +466,7 @@ private void reflectManifestGeneration(IMavenProjectFacade facade, MojoExecution
 	}
 
 	private void writeManifest(File manifestFile, Object manifest) throws UnsupportedEncodingException,
-			FileNotFoundException, IllegalAccessException, IllegalArgumentException, InvocationTargetException {
+			FileNotFoundException, IllegalAccessException, IllegalArgumentException, InvocationTargetException, IOException {
```
Comment (laeubi) [inline org.eclipse.m2e.mavenarchiver/src/org/eclipse/m2e/mavenarchiver/internal/AbstractMavenArchiverConfigurator.java:469]:
```suggestion
			IllegalAccessException, IllegalArgumentException, InvocationTargetException, IOException {
```

**Comment by laeubi** [comment]
> Thank for the info. I somehow managed to make it compile by [b582ab6](https://github.com/eclipse-m2e/m2e-core/commit/b582ab6644778045d15ae1a77b8bea8c9822ea03)
> 
> > we need to wait until most plugins have adopted to it
> 
> Do we have a list of these plugin and status of their compatibility with 3.9.0? Let me know if there's anything I can help.

3.9.0 is out for a few days so I won't expect much testing here, but best is to test different projects with 3.9.0 and report issues to the plugin authors.

**Comment by Eskibear** [comment]
Got it. Then let's wait a longer time for it... In the meanwhile, I'll do some tests as suggested.

**Comment by laeubi** [comment]
If you like just go on in fixing thing in m2e itself (so we are ready for 3.9.x) this already causes some test failures, so we probably need more adjustments here.

**Comment by Eskibear** [comment]
Most of the failures are caused by missing class `org.aopalliance.intercept.MethodInterceptor` from artifact `aopalliance:aopalliance`. It is excluded thus not packaged into the maven.runtime jar.

Lacking the context information ,I'm not sure why it's excluded historically, and whether it will work after we include it. I can give it a try and dig deeper anyway.

https://github.com/eclipse-m2e/m2e-core/blob/0b49dca241794f51aa31cd2e296d3184c564842e/m2e-maven-runtime/org.eclipse.m2e.maven.runtime/pom.xml#L55-L58

**Comment by laeubi** [comment]
Usually we exclude all what is not strictly required to keep the downloadsize small, if it is now a requirement it could be included.

**Comment by Eskibear** [comment]
> 19 failures / -60

By including aopalliance, we got 60 failed cases passed. The remaining 19 are:
- **maven-resources-plugin related** (15)
  - `org.eclipse.core.internal.resources.ResourceException: Errors occurred during the build.` (2)
    - [org.eclipse.m2e.tests.MarkerTest.testBuildContextWithTwoProjectConfigurators](https://ci.eclipse.org/m2e/job/m2e/job/PR-1238/3/testReport/junit/org.eclipse.m2e.tests/MarkerTest/testBuildContextWithTwoProjectConfigurators/)
    - ...
  - `Could not calculate build plan: Plugin org.apache.maven.plugins:maven-resources-plugin:3.3.0 or one of its dependencies could not be resolved: org.apache.maven.plugins:maven-resources-plugin:jar:3.3.0 was not found in file:repositories/remoterepo during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of central has elapsed or updates are forced` (13)
    - [org.eclipse.m2e.tests.conversion.ProjectConversionTest.testMultipleSourcesProjectConversion](https://ci.eclipse.org/m2e/job/m2e/job/PR-1238/3/testReport/junit/org.eclipse.m2e.tests.conversion/ProjectConversionTest/testMultipleSourcesProjectConversion/)
    - ...
- Build timed out. (2)
  - [org.eclipse.m2e.core.ui.tests.ConsoleTest.testConsole_debuggerAttachmentAndLinkAlignmentAndBehavior_mavenProject](https://ci.eclipse.org/m2e/job/m2e/job/PR-1238/3/testReport/junit/org.eclipse.m2e.core.ui.tests/ConsoleTest/testConsole_debuggerAttachmentAndLinkAlignmentAndBehavior_mavenProject/)
  - [org.eclipse.m2e.core.ui.tests.ConsoleTest.testConsole_debuggerAttachmentAndLinkAlignmentAndBehavior_tychoProject](https://ci.eclipse.org/m2e/job/m2e/job/PR-1238/3/testReport/junit/org.eclipse.m2e.core.ui.tests/ConsoleTest/testConsole_debuggerAttachmentAndLinkAlignmentAndBehavior_tychoProject/)
- Others (2)
  - [org.eclipse.m2e.core.ui.tests.ConsoleTest.testConsole_debuggerAttachmentAndLinkAlignmentAndBehavior_withColoredPrintout](https://ci.eclipse.org/m2e/job/m2e/job/PR-1238/3/testReport/junit/org.eclipse.m2e.core.ui.tests/ConsoleTest/testConsole_debuggerAttachmentAndLinkAlignmentAndBehavior_withColoredPrintout/)
  - [org.eclipse.m2e.tests.MarkerTest.testManagedTransitive](https://ci.eclipse.org/m2e/job/m2e/job/PR-1238/3/testReport/junit/org.eclipse.m2e.tests/MarkerTest/testManagedTransitive/)

**Comment by HannesWell** [comment]
Maven 3.9.1 has been released. So you probably want to update to that version directly.

**Comment by Eskibear** [comment]
Thanks for the info. 
I noticed there are regression bugs reported with 3.9.0, so I have been waiting for 3.9.1, as I was not sure whether above [build errors](https://github.com/eclipse-m2e/m2e-core/pull/1238#issuecomment-1422548105) would be fixed automatically. 
Honestly I have no clues about how to fix those "maven-resources-plugin related" errors, which is likely related to some caching mechanism. Maybe it's related the `RepositorySystem` instance after I see maven [3.9.1 release notes](https://github.com/apache/maven/releases/tag/maven-3.9.1).

**Comment by Eskibear** [comment]
> org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.eclipse.tycho:tycho-baseline-plugin:4.0.0-SNAPSHOT:verify (baseline-check) on project org.eclipse.m2e.core.ui: Baseline problems found! Project version: 2.0.4.20230305-1928, baseline version: 2.0.4.20230305-1928, suggested version: 2.0.5

tycho-baseline-plugin is unhappy, but I don't think I should update all related versions in this PR just for that. I remember previously it's skipped in Jenkins CI build. Any idea?

**Comment by laeubi** [comment]
It seems that there are some bytecode changes in the compiled code that now are getting visible, we are currently looking into this!

**Comment by laeubi** [comment]
@Eskibear if you rebase your branch the baseline problem should be fixed now!

**Comment by Eskibear** [comment]
Thanks, I'll revert 67b3cf2 and try rebase. 

BTW, I'm looking into below error:
> Could not calculate build plan: Plugin org.apache.maven.plugins:maven-resources-plugin:3.3.0 or one of its dependencies could not be resolved: The following artifacts could not be resolved: org.apache.maven.plugins:maven-resources-plugin:jar:3.3.0 (absent): org.apache.maven.plugins:maven-resources-plugin:jar:3.3.0 was not found in file:repositories/remoterepo during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of central has elapsed or updates are forced

I find lots of errors are related to missing maven-resources-plugin:3.3.0, and it's indeed not existing in https://github.com/tesla/m2e-core-tests/tree/c3acc1aa00a37eb9bdcebe8ad88d29233e1a4c71/org.eclipse.m2e.tests/repositories/remoterepo/org/apache/maven/plugins/maven-resources-plugin
Am I supposed to update `org.eclipse.m2e.tests/repositories/remoterepo` first?

**Comment by HannesWell** [comment]
> Am I supposed to update org.eclipse.m2e.tests/repositories/remoterepo first?

Maybe we should just add Maven-Central as another repo?
Given it's reliability I think it is OK to rely in it in tests. 
I contemplate to just remove all artifacts from such repo from git that are also available in Central.

**Comment by Eskibear** [comment]
> I contemplate to just remove all artifacts from such repo from git that are also available in Central.

I agree. That would make it much easier for new contributors who don't have a whole picture of the repo yet. 
Now it's somehow like a chicken-and-egg situation for me. In order to fix the CI build for this PR, I have to push maven-resources-plugin:3.3.0 into `m2e-core-tests` repo first. But in order to prove the push is necessary, it's better to get the CI build pass by adding the artifact.

**Comment by HannesWell** [comment]
You can prepare and open a PR for the tests submodule from your fork and then temporarliy adjust the .gitmodules file in this repo to point to the branch in your fork.
This way we can solve the Chicken-egg Problem and build all changes together in the CI.

**Comment by Eskibear** [comment]
@laeubi After the rebase, it seems the baseline problem still exists.

> Failed to execute goal org.eclipse.tycho:tycho-baseline-plugin:4.0.0-SNAPSHOT:verify (baseline-check) on project org.eclipse.m2e.core: Baseline problems found! Project version: 2.0.7.20230307-1553, baseline version: 2.0.7.20230307-1553, suggested version: 2.0.8 -> [Help 1]

logs from https://ci.eclipse.org/m2e/job/m2e/job/PR-1238/37/console

**Comment by laeubi** [comment]
This class extends `WorkspaceReader` what is a maven class so it is possible this has really changed and you can simply use the suggested version increment.

**Comment by Eskibear** [comment]
The baseline issue is gone, after I updated some versions following the plugin's suggestion.

As for the missing maven-resources-plugin:3.3.0 in m2e-core-tests, I followed https://github.com/eclipse-m2e/m2e-core/pull/1238#issuecomment-1475914068 with a commit https://github.com/Eskibear/m2e-core-tests/commit/93d053c751aa117188a756fb06f6ff8dbf86d8fb , adding .jar and .pom manually. Now it did found the artifact, but reporting below error messages:
> Could not calculate build plan: Plugin org.apache.maven.plugins:maven-resources-plugin:3.3.0 or one of its dependencies could not be resolved: Failed to read artifact descriptor for org.apache.maven.plugins:maven-resources-plugin:jar:3.3.0

I'm wondering if it's because parent POM & dependencies should also be added. BTW, I browsed some commits in m2e-core-tests, finding .pom and .jar files are usually added in batch. Two questions:
- Is there any scripts/plugins to add required artifacts? Or any pointer/clues to fix these errors?
- Besides running integration tests with `mvn verify -Pits`, any convenient way to debug a particular test case? (already set up the environment in Eclipse following the contributing guide)

**Comment by Eskibear** [comment]
> Test Results (13 failures / -7)

A good news is, after adding some missing artifacts in m2e-core-tests, it got 7 more cases passed. I'm now looking into the rest failures. Hopefully we can get them fixed soon. Thank you for your help!

**Comment by Eskibear** [comment]
> 3 failures (-10) , 7 skipped (±0)

Failures related to missing artifacts are all gone! 
See https://ci.eclipse.org/m2e/job/m2e/job/PR-1238/46/testReport/

Remaining 3 failures are all related to `org.eclipse.m2e.core.ui.tests.ConsoleTest.testConsole_xxx`, with below messages:
- `Build timed out.`
- `expected:<[simple.&#27;[1mSimpleTest]> but was:<[ForkStarter.java:282]>`

With a glance, I assume it's because new version of the related plugins generate console output in different format. Thus it failed to determine the finish by previous logic:
https://github.com/eclipse-m2e/m2e-core/blob/5c35b2329209e20f2a6ed1100f073d57b0b7542f/org.eclipse.m2e.core.ui.tests/src/org/eclipse/m2e/core/ui/tests/ConsoleTest.java#L337-L341

I'll give it a try, updating test cases to work with these new maven plugins.

**Comment by HannesWell** [comment]
Great progress. :)

> I checked raw output of mvn 3.8.6 and 3.9.1, the finishing output message are the same, including the ansi color escape sequences.
> 
> I have no idea how to investigate the failures related to `ui.tests.ConsoleTest`, mentioned in [#1238 (comment)](https://github.com/eclipse-m2e/m2e-core/pull/1238#issuecomment-1477224251)
> 
> @HannesWell From the commit history I find you might be more familiar with this part. Can you spare some time take a look at it? Great appreaciation!!!
> 

Yes I will look into this as soon as possible, I'll try it tomorrow but cannot promise this.

> BTW, after you rebased the commits, integration tests in Jenkins CI failed to build due to below error, while ITs in GitHub Action is ok. I'm wondering if it's due to some cached snapshot artifacts created in previous runs...
> 
> > 16:59:12  [ERROR] Failed to execute goal org.eclipse.tycho:tycho-p2-plugin:4.0.0-SNAPSHOT:p2-metadata (p2-metadata) on project org.eclipse.m2e.core: baseline and build artifacts have same version but different contents
> > 16:59:12  [ERROR]    no-classifier: different
> > 16:59:12  [ERROR]       org/eclipse/m2e/core/internal/project/registry/EclipseWorkspaceArtifactRepository.class: different

Yes that because now there is a change in the resulting jar without a change in the git-timestamp. The GH-Workflows are setup less strict in this regard because Jenkins is the one that also builds the baseline.
To make it short, you need a new git-timestamp in this project, so have to modify any file in the project. If there is nothing reasonable to change we usually use the `forceQualifierUpdate.txt` file and write a little note about the reason in it.

**Comment by Eskibear** [comment]
@HannesWell Thank you for the information. I also find some clues when I look into `org.eclipse.m2e.core.ui.tests.ConsoleTest`. Hope it helps.

- (1) testConsole_debuggerAttachmentAndLinkAlignmentAndBehavior_mavenProject
  - **Build timed out.**
  - 2 min 4 sec
- (2) testConsole_debuggerAttachmentAndLinkAlignmentAndBehavior_withColoredPrintout
  - expected:<[simple.&#27;[1mSimpleTest]> but was:<[**ForkStarter.java:282**]>
  - 7 sec
- (3) testConsole_debuggerAttachmentAndLinkAlignmentAndBehavior_tychoProject
  - **Build timed out**.
  - 2 min 3 sec

I downloaded the IDE built from the JenkinsCI, and ran `mvn clean verify -Dmaven.surefire.debug=true` manually according to the test cases. 

In ITs, above test cases are executed sequentially. 
#### Test (1) stuck on listening 5005, but no debugger was attaching. 
That's why the build timed out after 120s, failed to find `[INFO] Finished at` in output.
```
[INFO] Listening for transport dt_socket at address: 5005
```
#### Test (2) failed because port 5005 was still in use.
And the output was like below:
image
And that's why the link text is `ForkStarter.java:282`
#### Test (3) continued to timeout, because it's using a new port 8000, same as the behavior in (1). 
https://github.com/eclipse-m2e/m2e-core/blob/270e3569672b5a31572dc671f625bcc580987828/org.eclipse.m2e.core.ui.tests/src/org/eclipse/m2e/core/ui/tests/ConsoleTest.java#L168-L173

What confuses me is, I used both mvn 3.9.1 and 3.8.7 to run the command, and got the same behavior, why the test cases only failed for this PR? Is any debugger expected to attach to the process during the tests?

**Comment by HannesWell** [comment]
There are multiple issues that prevent the ConsoleTest from succeeding.

First the project description printed on the console in the beginning of the build of each project changed from something like

```
[INFO] --------< org.eclipse.m2e.tests:simple.projectWithJUnit-5_Test >--------
[INFO] Building simple.projectWithJUnit-5_Test 1
[INFO] --------------------------------[ jar ]---------------------------------
```
to
```
[INFO] --------< org.eclipse.m2e.tests:simple.projectWithJUnit-5_Test >--------
[INFO] Building simple.projectWithJUnit-5_Test 1
[INFO]   from pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
```
And therefore the format expected by the `MavenConsoleLineTracker` was wrong and the tracker never entered the 'in-project-build' mode. And when the tracker does not know which project is currently build it cannot launch a debugger.
OK, that was not too hard to fix, I pushed another commit on your branch.

The second thing that changed is that since version 3.0.0-M6 the maven-surefire-plugin does not print the `Listening for transport` line anymore, but the `MavenConsoleLineTracker` waits for that line to extract the port to which it can launch a debugger.

Before it looked like

```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Listening for transport dt_socket at address: 5005
[INFO] Running simple.SimpleTest
```
and now it is just
```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running simple.SimpleTest
```
This site seem to explain the reason:
https://maven.apache.org/surefire/maven-surefire-plugin/examples/process-communication.html
And when I add the following configuration to the surefire-plugin the debugger successfully attaches to the test-process.
```
<configuration>
	<forkNode implementation="org.apache.maven.plugin.surefire.extensions.SurefireForkNodeFactory"/>
</configuration>
```

So to fix this test for now we have to add that configuration option (or revert back to surefire 2.22.2).
But on the long run we have think about a more robust solution how to detect a forked process that needs a debugger, that does not rely on a certain print-out.
Maybe something that also completes https://github.com/eclipse-m2e/m2e-core/pull/471. I'm currently thinking about either to poll the process descendants regularly (a notification would be nicer) or to introduce an agent that manipluates the byte code and intercepts the launching of processes.

**Comment by Eskibear** [comment]
@HannesWell  Thank you for the help!!!

> ... a more robust solution how to detect a forked process that needs a debugger, that does not rely on a certain print-out.

Totally agree, especially due to potential changes like this in output format, which is not guaranteed to stay the same.

> I'm currently thinking about either to poll the process descendants regularly (a notification would be nicer) 

It looks promising, polling with some reasonable timeout would be simpe and effective. Notification mechanism is better but might have requirements on upstream projects like surefire plugin.

> or to introduce an agent that manipluates the byte code and intercepts the launching of processes.

This looks too hacky to me and the processes are not launched in a common way that we want to guard?

Above are just my two cents given my limited knowledge about m2e, maybe further discussion can happen in another issue like #471 .

And again, thank you Hannes for fixing test failures and explainig the root cause underneath! Now I think this PR should be ready for review.
(In GH workflow all tests pass, but Jenkins CI has one new failure related to lemminx, maybe unrelated?)

**Comment by HannesWell** [comment]
> > or to introduce an agent that manipluates the byte code and intercepts the launching of processes.
> 
> This looks too hacky to me and the processes are not launched in a common way that we want to guard?

It is indeed hacky, but the main idea is to also be able to inject jdwp arguments in forked java processes, so that M2E can control if forked java-process will wait for a debugger or not, without the need to have a mapping for the configuration of each Plugin.
This implicitly also gives a reliable mechanism for notifications about launches of forked processes. 
But yes this is not yet decided and worked out and there are also other suggestions.



> And again, thank you Hannes for fixing test failures and explainig the root cause underneath! Now I think this PR should be ready for review.
> (In GH workflow all tests pass, but Jenkins CI has one new failure related to lemminx, maybe unrelated?)

You are very welcome and thanks for your work and your patience as well.
I'm not so sure if that failure is really unrelated. Locally in my Dev-Eclipse it succeeds, but even after two re-builds the failure is still there and the master succeeds.
Maybe @mickaelistria do you have an idea?

Btw. I created a competing PR for the m2e-core-tests repo to remove all artifacts in the embedded `remoterepo`:
https://github.com/tesla/m2e-core-tests/pull/200

But this could be a lengthy task, so I'm not sure if this is worth it. If I don't succeed over the weekend I will merge your PR https://github.com/tesla/m2e-core-tests/pull/199 so that we can proceed here.

But I think we are close to the finish line here. :)

**Comment by mickaelistria** [comment]
The lemminx failure is most likely related. I will try to reproduce it locally soon. I'll prioritize it to allow more progress here. Please ping me in the beginning of next week if I didn't give news on this topic, as I may have forgotten about it over the week-end.

**Comment by Eskibear** [comment]
I cannot reproduce the lemminx failure either. Let me rebase it on top of the master branch.

**Comment by Eskibear** [comment]
> If this works for you as well, I think we can just merge it and assume the failure is just "bad luck" and cross fingers so it doesn't happen on master.

Just a reminder that before merging this PR, we probably need to merge https://github.com/tesla/m2e-core-tests/pull/199 first and update submodule.

**Comment by HannesWell** [comment]
Now that #1130 is merged I rebased this PR on top of the current master and squashed all commits and added myself as co-auther. Feel free to add some more comments to the commit message if you think it is necessary.
In my previous attempt with the deleted embedded remote repo there were two test-failures in:
- org.eclipse.m2e.tests.MarkerTest.testBuildContextWithSameProjectConfiguratorTwice
- org.eclipse.m2e.editor.lemminx.tests.EditorTest.testOpenChildThenParentResolvesParent

From the latter we know it occasionally fails and the latter occured before and you fixed it with adding more resources to the embedded remoterepo. However I was not able to reproduce it locally, so lets see if this was by chance as well.
Once the build is green I think this is ready for submission.

**Comment by Eskibear** [comment]
Feel free to merge it. Thank you!

**Comment by Eskibear** [comment]
Forgot to mention: 

> [org.eclipse.m2e.tests.MarkerTest.testBuildContextWithSameProjectConfiguratorTwice](https://ci.eclipse.org/m2e/job/m2e/job/PR-1238/59/testReport/org.eclipse.m2e.tests/MarkerTest/testBuildContextWithSameProjectConfiguratorTwice/)

This requires an update of the test case as https://github.com/tesla/m2e-core-tests/pull/199/commits/49e2b2241abf10df6402fa98a7e388893acb75f9

**Review by laeubi — APPROVED** [review]

**Comment by Eskibear** [comment]
For your convenience, I create https://github.com/tesla/m2e-core-tests/pull/201 , which I believe can also be merged at the time you merge this PR.

**Review by HannesWell — APPROVED** [review]
Thanks a lot @Eskibear for this non trivial contribution and your patience.
This is highly appreciated.

**Pull request merged**